texts/0004-mobile-plugin.md

```diff
@@ -0,0 +1,201 @@
+- Feature Name: mobile_plugin
+- Start Date: 2023-03-23
+- RFC PR: [tauri-apps/rfcs#0004](https://github.com/tauri-apps/rfcs/pull/9)
+
+# Summary
+
+Tauri needs access to native mobile APIs such as camera, geolocation and notifications. These APIs can be called by Rust code using `jni-rs` and `objc`, but writing Kotlin or Java code for Android and Swift code for iOS simplifies development and is a much more mature way of accessing these APIs.
+
+# Motivation
+
+With access to native Android projects and Swift packages for iOS, Tauri and its users can use any mobile API a native application might need. This proposal includes a way to call the native mobile code from both the webview and the Rust layers, giving more control to both application and library writers.
+
+# Guide-level explanation
+
+For the sake of simplicity, we can simply extend the existing Tauri plugin API to allow registering a native mobile project and linking it to the application.
+
+## Build Script
+
+The plugin's build script will be responsible for injecting the plugin's Android project to the application's project inside the `gen/android/.tauri/plugins` folder, and compiling and linking the Swift library for iOS using `swift-rs`:
+
+```rust
+use std::process::exit;
+
+fn main() {
+  if let Err(error) = tauri_build::mobile::PluginBuilder::new()
+    .android_path("android")
+    .ios_path("ios")
+    .run()
+  {
+    println!("{error:#}");
+    exit(1);
+  }
+}
+
+```
+
+## Example Plugin
+
+Let's see an example plugin code for both Android and iOS:
+
+```kotlin
+package com.plugin.example
+
+import android.app.Activity
+import app.tauri.annotation.Command
```

Comment:
Is this going to add Swift and Kotlin Tauro APIs we need to document as well? Do we know if there are websites we can point to that would generate and host those or would we have to build it ourselves on the website?

Reply:
We do have an Android and a Swift library the plugin writers are going to use. I can research about documentation tomorrow.

Reply:
DocC build script and output:

```zsh
xcrun --find docc
mkdir Tauri.docc
alias swiftios="swift build -Xswiftc -sdk -Xswiftc /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS16.2.sdk -Xswiftc -target -Xswiftc arm64-apple-ios13.0"
swiftios -Xswiftc -emit-symbol-graph -Xswiftc -emit-symbol-graph-dir -Xswiftc .build
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/docc preview Tauri.docc --fallback-display-name Tauri --fallback-bundle-identifier app.tauri --fallback-bundle-version 1 --additional-symbol-graph-dir .build
```

<img width="230" alt="Screenshot 2023-03-30 at 13 04 30" src="https://user-images.githubusercontent.com/20051258/228897191-e232fa91-e243-4869-8080-e046826edfae.png">

Reply:
Looks like an actual Apple documentation page :)
<img width="1512" alt="Screenshot 2023-03-30 at 13 07 09" src="https://user-images.githubusercontent.com/20051258/228897420-3806d41b-7804-4c37-bb4b-e7c1fcfcede1.png">

Reply:
For the Android documentation we'll need to use [KDoc](https://kotlinlang.org/docs/kotlin-doc.html).